catlas/cache_utils.py

```diff
@@ -99,3 +125,155 @@ def check_cache(cached_func):
         return True
 
     return False
+
+
+NO_CACHE_RESULT = "not a cache result"
+
+
+def sqlitedict_memoize(
+    folder,
+    func,
+    ignore=(),
+    mmap_mode=None,
+    shard_digits=2,
+):
+
+    # Use a base name that includes the full function path and a hash on the function code itself
+    base = better_build_func_identifier(func)
+
+    @backoff.on_exception(
+        backoff.expo, (sqlite3.OperationalError, TimeoutError), max_time=180
+    )
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        """Wrapper for callable to cache arguments and return values."""
+
+        # Get the key for the arguments
+        key = hash(
+            filter_args(func, ignore, args, kwargs), coerce_mmap=(mmap_mode is not None)
+        ).encode("utf-8")
+
+        # Generate a base db location based on the full function path and a hash on the function code itself
+        db_loc = ".".join(base)
+
+        # Add a couple of shard digits so that actually we reference a subfolder (string is hex, so two shard digits is 16^2 shards)
+        # the numer of shards should be approximately equal to the number of expected simultaneous writers
+        if shard_digits > 0:
+            db_loc = str(db_loc) + "/" + key.decode()[0:shard_digits] + ".sqlite"
+
+        result = NO_CACHE_RESULT
+
+        cache = SqliteSingleThreadDict(
+            folder + db_loc,
+        )
+
+        # Grab the cached entry (might be None)
+        # if key in cache:
+        try:
+            cache.__enter__()
+            result = cache[key]
+        except (pickle.PicklingError, sqlite3.OperationalError, KeyError) as e:
+            # It's in the cache, but the pickle data is corrupted :(
+            pass
+
+        # If we need to compute the result, do it, but make sure we only do this once
+        # in case we're hitting it due to cache connection problems.
+        if (type(result) == str) & (result == NO_CACHE_RESULT):
+            result = func(*args, **kwargs)
+            cache[key] = result
+
+        cache.close()
+
+        del cache
+        return result
+
+    return wrapper
+
+
+class SqliteSingleThreadDict(dict):
+    def __init__(
+        self,
+        filename=None,
+        tablename="unnamed",
+        journal_mode="PERSIST",
+        encode=cloudpickle.dumps,
+        decode=cloudpickle.loads,
+    ):
+        self.filename = filename
+        self.journal_mode = journal_mode
+        self.encode = encode
+        self.decode = decode
+        self.tablename = tablename
+
+    def _new_ro_conn(self):
```

Comment:
definitely a nitpick but it would be nice if this function was more human readable. What is ro?